### PR TITLE
Xgboost

### PR DESCRIPTION
**Pull request by moreira-presh**
This is a link to the notebook  [here](https://github.com/HamoyeHQ/g01-fraud-detection/blob/c9bcd0825b28fa2c9f266a71ddfb42c56e0b3b99/model/XGBOOST%20-%20fraud%20detection.ipynb)

I had problems incorporating the cost saving function, and using grid search too forever.

### CODE REVIEW

**Comment by moreira-presh** [comment]
Please go through the notebook for added features and all

**Comment by Jolomi-Tosanwumi** [comment]
Well done @moreira-presh ...however, you need to set shufffle=True in the StratifiedKFold for the first try.
Also, ROC AUC is not a good metric for imbalanced dataset as it is misleading. Use f1score and/or AUC precision-recall implemented as average_precision_score in sklearn.

**Comment by Jolomi-Tosanwumi** [comment]
Also, your downsampling ratio (from 298k to 600) for the majority class may be too over biased and not generalize well. Especially for problem as fraud detection; too much down sampling may only give empirical results but not generalized results. You may want to increase the downsampling ratio and use other non-biased downsampling llike tomi links or nearest neighbors. Although, downsampling may not be a good choice for fraud detection because a fraud detecting model is required to detect each and every transaction. Downsampling may not answer the question on what a model thinks about the removed observations...i.e what can it predict them as. It also lead to loss of information. Emperically, as we reduce the majority class, the model f1 score will increase. It is an emperical theory and may not tell well of a model's performance.

**Comment by Jolomi-Tosanwumi** [comment]
You may also want to stratify your train_test_split. You may also want to use Standard Scaler over MinMaxScaler because of the reasons I explain in our Slack channel.

**Comment by Jolomi-Tosanwumi** [comment]
Hi @moreira-presh ...you may also want to avoid data leakage by doing train_test_split first before undesampling the training set. Or you encapsulate scaler, undersampler and the model in an imblearn pipeline. This is because I see data leakage due to the fact that you undersampled before splitting. This may also give misleading results. You may want to increase the undersampling ratio, then split before undersampling. 

Hope this helps.

Thank you.

**Comment by moreira-presh** [comment]
Yes would follow suit

**Comment by moreira-presh** [comment]
Greetings bro..
Nice work you are doing, I must commend :raised_hands:.Not untill now I didn't realize how dealing with imbalanced dataset problem was important,
Working on my part of the to-do (XG Boost)
I tried correcting some of the previous work I did on my model:
1. Just working on it without scaling, sampling or doing any cross val. Had accuracy of 1, which is obviously overfitting, the F-scofe was also around 0.81 + or so, and seemed good :smirk: 2. Carrying out a sample ratio of 6000 (undersampling), standard scaling, [stratified during train test split although it was done before sampling]. The accuracy was 0.98, AUC = 0.981, avg_precision_score = 0.836 and f1 = 0.91.I still think there's a problem with it,
1:27
I would do the sampling on the train set after splitting the data, which is the right thing to do...
Pls your suggestions would be appreciated..I'm thinking of applying weights=balanced for penalizing,
